IOP message box send/recieve tests

jpd002 opened the pull request:
Basic tests for SendMbx and ReceiveMbx functions on the IOP. Tests all queuing modes and some invalid parameters.

jpd002 committed:
Added IOP message box send test.

jpd002 committed:
Moved common message box functions to a common file.

Common functions will be used in other tests.

jpd002 committed:
Added IOP message box receive test.

unknownbrackets commented on tests/kernel/iop/messagebox/mbx-common.c:
```diff
@@ -0,0 +1,32 @@
+
+void printMbx(s32 mbxId) {
+	iop_mbx_status_t mbxInfo;
+	s32 result = ReferMbxStatus(mbxId, &mbxInfo);
```
Do the "reserved" values have anything interesting?  I've found that so-called reserved values can often be important, at least in pspsdk.

Also, any reason to log the attr/option values, just in case they vary?

-[Unknown]

jpd002 replied:
Oh sorry, I've completely missed this! No idea if the reserved values contain anything interesting, but I'll take a look. And yeah, attr and option values should definitively be logged.

unknownbrackets commented on tests/kernel/iop/messagebox/receive.c:
```diff
@@ -0,0 +1,87 @@
+	printf("invalid params:\n");
+	{
+		int result = ReceiveMbx(NULL, ~0);
+		printf("  receiving message from invalid box: %d\n", result);
```
Does it crash if NULL is used as the destination with a valid mbox?

Also, if it errors out, does it leave the destination unchanged (e.g. even if it wasn't previously null)?  Presumably yes.

-[Unknown]

unknownbrackets commented on tests/kernel/iop/messagebox/receive.c:
```diff
@@ -0,0 +1,87 @@
+	testMultipleReceive(MBA_THFIFO | MBA_MSFIFO);
+
+	printf("receiving (priority thread queuing):\n");
+	testMultipleReceive(MBA_THPRI | MBA_MSFIFO);
```
It's probably worth testing MBA_MSPRI in another test.  I'm guessing the priority field is ignored without it?

Also would be interesting to validate that the "unused" bits outside priority are in fact unused, by comparing them with 0xFF or 0x01 in them with 0x00.  Just in case the priority isn't treated as a u8 after all.

-[Unknown]

jpd002 replied:
Message priority queuing is tested in SendMbx test. I thought it would make more sense to put it there because it's basically a "feature" of that function even though the mode is set on the message box itself.

Hadn't thought of testing the "unused" fields, but it might be interesting to test those at some point.

unknownbrackets replied:
Ah, I missed that they weren't both thread.  Cool.

-[Unknown]

unknownbrackets commented on tests/kernel/iop/messagebox/receive.c:
```diff
@@ -0,0 +1,87 @@
+	schedf("  order: ");
+	for(i = 0; i < msgCount; i++) {
+		memset(&msg[i], 0, sizeof(MSG));
+		msg[i].header.priority = msgCount - i;
```
Another test, but would be interesting if there's any validation here on a max msg priority.

-[Unknown]

unknownbrackets commented on tests/kernel/iop/messagebox/receive.c:
```diff
@@ -0,0 +1,87 @@
+#define STACK_SIZE 0x800
+
+typedef struct
+{
```
Nit: it's weird to see structs as the only thing using a different bracing style.  Doesn't really matter though.

-[Unknown]

unknownbrackets commented on tests/kernel/iop/messagebox/send.c:
```diff
@@ -0,0 +1,92 @@
+	printf("invalid params:\n");
+	{
+		int result = SendMbx(~0, NULL);
+		printf("  sending message to invalid box: %d\n", result);
```
It's sometimes interesting to delete a box and then try sending or receiving.  On the PSP, this generally fails, but there are some cases where it doesn't and continues to return data from structs used by the kernel (so still "correct".)  Sometimes PSP games will have bugs where they rely on it working this way, even when logically it shoudln't.

-[Unknown]

jpd002 replied:
Woah, that looks quite scary. Is it because the message box isn't really deleted when sceKernelDeleteMbx is called (as if it had some kind of "zombie" state)?

unknownbrackets replied:
Well, it wasn't messsage boxes.  I don't recall off the top of my head which kernel object it was... maybe LwMutexes?  Hmm.

But, for example, sceAtrac has a function that allows you to get the "context" which is a writable struct of state for playing music.  In general it's only used by a few games to integrate with another sound playing library (which simulates the PS1/PS2 SPU, I think), but this struct will actually persist even after the atrac object is deleted - because it is a static array in memory.

There were a couple games that relied on memory allocation order in an FPL (block-based allocator), since they would double-free all the time, and it only happened to work when the allocation order was round-robin (which is what the PSP does for FPLs.)

-[Unknown]

jpd002 replied:
Interesting. I wonder if the IOP's FPL exhibits the same behavior as the PSP's.

unknownbrackets commented on tests/kernel/iop/messagebox/send.c:
```diff
@@ -0,0 +1,92 @@
+		printf("  sending message to invalid box: %d\n", result);
+
+		//Sending a null message does seem to work even though it probably has
+		//dangerous side effects since the message header at address 0 will be modified
```
Ouch.  This is why null should never be a mapped address.  Glad it isn't on the PSP.  Thanks for keeping this noted in a comment.

-[Unknown]

unknownbrackets commented on tests/kernel/iop/messagebox/send.c:
```diff
@@ -0,0 +1,92 @@
+		memset(&msg[i], 0, sizeof(MSG));
+		msg[i].header.priority = msgCount - i;
+		msg[i].payload         = i + 1;
+		SendMbx(mbxId, &msg[i]);
```
Does anything interesting happen if the `next` value already points to another message?  For example, can I send two messages at once?  I'm assuming it won't work, but I wonder if it will actually clear the `next` value to null or just leave it.  That might be important behavior.

-[Unknown]

unknownbrackets commented:
Looks good.  I noted a few other areas that we could test but if you'd prefer to just merge as is, I'm cool with that.  Already a good test.

-[Unknown]

jpd002 committed:
Code style fixes.

jpd002 committed:
Added missing DeleteMbx call.

jpd002 committed:
Added comment about using ReceiveMbx with a null pointer.

jpd002 committed:
Added SendMbx message linking behavior test.

jpd002 commented:
Thanks for the review. 

I've fixed the various code style issues, added a comment about ReceiveMbx working even though a NULL pointer is passed and added a test to verify how the next pointer in the header is modified by SendMbx.

That's pretty much what I wanted to test for now, I'll add other tests cases if I come across some weird behavior when debugging games in the emulator.
